### PR TITLE
feat: new dsl for native binaries

### PR DESCRIPTION
**Pull request by danil-pavlov**

### CODE REVIEW

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
+[//]: # (title: Build Kotlin/Native artifacts)
+
+> The new DSL described below is [Experimental](components-stability.md). It may be changed at any time.
+> We encourage you to use it for evaluation purposes.
+> 
+{type="warning"}
+
+Kotlin/Native [targets](multiplatform-dsl-reference.md#native-targets) are compiled to the `*.klib` library artifacts,
+which can be consumed by Kotlin/Native itself as a dependency but cannot be executed or used as a native library.
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:9]: The new DSL is only for libraries not for executables.
```
itself as a dependency but cannot be used as a native library.
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+ 
+To declare final native binaries such as executables or shared libraries, use the new binaries format with the `kotlinArtifacts`
+DSL. It represents a collection of native binaries built for this target in addition to the default `*.klib` artifact 
+and provides a set of methods for declaring and configuring them.
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:13]:
```
To declare final native libraries, use the new ...
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+| Factory method | Binary kind             | Available for                                |
+|----------------|-------------------------|----------------------------------------------|
+| `sharedLib`    | Shared native library   | All native targets, except for `WebAssembly` |
+| `staticLib`    | Static native library   | All native targets, except for `WebAssembly` |
+| `framework`    | Objective-C framework   | macOS, iOS, watchOS, and tvOS targets only   |
+| `fatFramework` | Universal fat framework | macOS, iOS, watchOS, and tvOS targets only   |
+| `XCFramework`  | XCFramework framework   | macOS, iOS, watchOS, and tvOS targets only   |
+
+The simplest version doesn't require any additional parameters and creates one binary for each build type. Currently,
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:32]: It requires either a target or targets indeed

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+```groovy
+kotlin {
+    kotlinArtifacts {
+        Native.Library {
+            target = iosX64 // Define your target instead
+            modes(DEBUG, RELEASE)
+            // Binary configuration
+        }
+    }
+}
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:68]:
```
kotlin {
    kotlinArtifacts {
        it.native.Library {
            target = iosX64 // Define your target instead
            modes(DEBUG, RELEASE)
            // Binary configuration
        }
    }
}
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+
+</tab>
+<tab title="Groovy" group-key="groovy">
+
+```groovy
+kotlinArtifacts {
+    Native.Library("mylib") {
+        // Binary configuration
+    }
+}
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:95]:
```
kotlinArtifacts {
    it.native.Library("mylib") {
        // Binary configuration
    }
}
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+```
+
+</tab>
+</tabs>
+
+The argument sets a name prefix, which is the default name for the binary file. For example, for Windows the code 
+produces the files `mylib.exe`.
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:102]: For example, for Windows the code produces the file `mylib.dll`.

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+
+## Build binaries for several modules
+
+If your project has multiple Kotlin modules, and you need to access them from your iOS app, you'll encounter an issue
+− the usage of several Kotlin/Native modules in Swift is limited.
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:107]: The problem exists for any native target not only for iOS

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+With Kotlin artifact DSL, you can export the project along with all its dependencies into a single Gradle module
+and connect it to Swift.
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:110]: With Kotlin artifact DSL, you can export multiple Kotlin modules into a single artifact.

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+
+</tab>
+<tab title="Groovy" group-key="groovy">
+
+```groovy
+kotlinArtifacts {
+    Native.Library("mylib") {
+        // Binary configuration
+    }
+}
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:175]:
```
kotlinArtifacts {
    it.native.Library("mylib") {
        target = linuxX64
        kotlinOptions {
            freeCompilerArgs += "-Xmen=pool"
        }
    }
    it.native.Library("myslib") {
        target = linuxX64
        it.static = false
        modes(DEBUG)
        addModule(project(":lib"))
        kotlinOptions {
            verbose = false
            freeCompilerArgs = []
        }
    }
}
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:217]:
```
kotlinArtifacts {
    it.native.Framework("myframe") {
        modes(DEBUG, RELEASE)
        target = iosArm64
        it.static = false
        embedBitcode = EmbedBitcodeMode.MARKER
        kotlinOptions {
            verbose = false
        }
    }
}
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+The registered Gradle task is `assembleMyslibSharedLibrary` that assembles all types of registered "myslib" into a dynamic library.
+
+#### Framework
+
+For the framework configuration, the following additional parameters are available:
+
+| **Name**       | **Description**                                                                                                                                        |
+|----------------|--------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `target`       | Declares a particular target of a project. The names of available targets are listed in the [Targets](multiplatform-dsl-reference.md#targets) section. |
+| `embedBitcode` | Declares the mode of bitcode embedding. Use `MARKER` to embed the bitcode marker (for debug builds) or `DISABLE` to turn off embedding.                |
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:190]: embedBitcode is not needed anymore with Xcode 14

Comment (terrakok) [reply docs/topics/multiplatform/multiplatform-native-artifacts.md:190]: but the option will be here for users anyway

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+
+</tab>
+<tab title="Groovy" group-key="groovy">
+
+```groovy
+kotlinArtifacts {
+    Native.Library("mylib") {
+        // Binary configuration
+    }
+}
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:266]:
```
kotlinArtifacts {
    it.native.FatFramework("myfatframe") {
        targets(iosX32, ios64)
        embedBitcode = EmbedBitcodeMode.DISABLE
        kotlinOptions {
            suppressWarnings = false
        }
    }
}
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+| `targets`       | All targets of the project.                                                                                                             |
+| `embedBitcode`  | Declares the mode of bitcode embedding. Use `MARKER` to embed the bitcode marker (for debug builds) or `DISABLE` to turn off embedding. |
+
+<tabs group="build-script">
+<tab title="Kotlin" group-key="kotlin">
+
+```kotlin
+kotlinArtifacts {
+    Native.FatFramework("myfatframe") {
+        targets(iosX32, ios64)
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:249]: iosX64

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+
+</tab>
+<tab title="Groovy" group-key="groovy">
+
+```groovy
+kotlinArtifacts {
+    Native.Library("mylib") {
+        // Binary configuration
+    }
+}
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:313]:
```
it.native.XCFramework("sdk") {
    targets(iosX64, iosArm64, iosSimulatorArm64)
    setModules(
        project(":shared"),
        project(":lib")
    )
}
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+* `podPublishDebugXCFramework`, which generates a debug XCFramework along with a podspec file.
+* `podPublishXCFramework`, which generates both debug and release XCFrameworks along with a podspec file.
+
+This can help you distribute shared parts of your project separately from mobile apps through CocoaPods. You can also use XCFrameworks
+for publishing to private or public podspec repositories.
+
+> Publishing Kotlin frameworks to public repositories is not recommended if those frameworks are built for different versions
+> of Kotlin. Doing so might lead to conflicts in the end-users' projects.
+>
+{type="warning"}
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:339]: delete

**Comment by terrakok** [comment]
It might be better:
 - for `library` and `framework` user has to setup `target`
 - for `fatframework` and `xcframework` user has to setup `targets`

For all Apple artifacts user can setup `embedBitcode` mode.

All other options are the same for all types of artifacts.

**Comment by terrakok** [comment]
code examples some time contain `kotlin {}` block some time not. I think this should be consistent

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+
+| Factory method | Binary kind             | Available for                                |
+|----------------|-------------------------|----------------------------------------------|
+| `sharedLib`    | Shared native library   | All native targets, except for `WebAssembly` |
+| `staticLib`    | Static native library   | All native targets, except for `WebAssembly` |
+| `framework`    | Objective-C framework   | macOS, iOS, watchOS, and tvOS targets only   |
+| `fatFramework` | Universal fat framework | macOS, iOS, watchOS, and tvOS targets only   |
+| `XCFramework`  | XCFramework framework   | macOS, iOS, watchOS, and tvOS targets only   |
+
+The simplest version requires the `target` (or `targets`) and `modes` parameters for the selected build type. Currently,
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:31]: Only target (or targets) is required.
Modes have default value equal `modes(DEBUG, RELEASE)`

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+        }
+    }
+}
+```
+
+</tab>
+</tabs>
+
+The argument sets a name prefix, which is the default name for the binary file. For example, for Windows the code 
+produces the files `mylib.dll`.
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:105]: `the file`

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+
+## Build binaries for several modules
+
+If your project has multiple Kotlin modules, and you need to access them from your iOS app, you'll encounter an issue
+− the usage of several Kotlin/Native modules in Swift is limited. With Kotlin artifact DSL, you can export multiple
+Kotlin modules into a single artifact.
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:111]: If your project has multiple Kotlin modules, and you need to access them from your iOS app, you'll encounter an issue − the usage of several Kotlin/Native **artifacts** in Swift is limited.

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+            modes(DEBUG)
+            addModule(project(":lib"))
+            kotlinOptions {
+                verbose = false
+                freeCompilerArgs = emptyList()
+            }
+        }
+    }
+}
+```
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:168]:
```kotlin
kotlinArtifacts {
    Native.Library("myslib") {
        target = linuxX64
        isStatic = false
        modes(DEBUG)
        addModule(project(":lib"))
        kotlinOptions {
            verbose = false
            freeCompilerArgs += "-Xmen=pool"
        }
    }
}
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+            modes(DEBUG)
+            addModule(project(":lib"))
+            kotlinOptions {
+                verbose = false
+                freeCompilerArgs = []
+            }
+        }
+    }
+}
+```
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:194]:
```groovy
kotlinArtifacts {
    it.native.Library("myslib") {
        target = linuxX64
        it.static = false
        modes(DEBUG)
        addModule(project(":lib"))
        kotlinOptions {
            verbose = false
            freeCompilerArgs += "-Xmen=pool"
        }
    }
}
```

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 21]
```diff
@@ -0,0 +1,370 @@
...
+
+</tab>
+</tabs>
+
+The registered Gradle task is `assembleMyframeFramework` that assembles all types of registered "myframe" framework.
+
+> If for some reason the new DSL doesn't work for you, try [this workaround](multiplatform-build-native-binaries.md#export-dependencies-to-binaries)
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:253]: instead `this workaround` I would use `old approach`

**Comment by terrakok** [comment]
For all code blocks need to delete `kotlin {}` wrapper

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+| `isStatic`      | Optional linking type that defines the library type. By default, it's `false` and the library is dynamic.                       |
+| `modes`         | Optional build types, `DEBUG` and `RELEASE`.                                                                                    |
+| `kotlinOptions` | Optional compiler options applied to the compilation. See the list of available [compiler options](gradle.md#compiler-options). |
+| `addModule`     | In addition to the current module, you can add other modules to the resulting artifact.                                         |
+| `setModules`    | You can override the list of all modules that will be added to the resulting artifact.                                          |
+
+### Libraries and frameworks
+
+When building an Objective-C framework or a native library (shared or static), you may need to pack not just the classes
+of the current project, but also the classes of its dependencies into a single entity and export all these modules to it.
```
Comment (terrakok) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:126]: `but also the classes of its dependencies` -> `but also the classes of any other multiplatform module`

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 23]
```diff
@@ -0,0 +1,335 @@
...
+
+By default, an Objective-C framework produced by Kotlin/Native supports only one platform. However, you can merge such
+frameworks into a single universal (fat) binary. This especially makes sense for 32-bit and 64-bit iOS frameworks.
+In this case, you can use the resulting universal framework on both 32-bit and 64-bit devices.
+
+For the fat framework configuration, the following additional parameters are available:
+
+| **Name**       | **Description**                                                                                                                                                                                   |
+|----------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `targets`      | All targets of the project.                                                                                                                                                                       |
```
Comment (koshachy) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:241]: Declares all targets of the project?

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 24]
```diff
@@ -0,0 +1,335 @@
...
+            suppressWarnings = false
+        }
+    }
+}
+```
+
+</tab>
+</tabs>
+
+The registered Gradle task is `assembleMyfatframeFatFramework` that assembles all types of registered "myfatframe" FatFramework.
```
Comment (koshachy) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:277]: "myfatframe" fat framework?

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 25]
```diff
@@ -0,0 +1,335 @@
...
+            project(":lib")
+        )
+    }
+}
+```
+
+</tab>
+</tabs>
+
+The registered Gradle task is `assembleSdkXCFramework` that assembles all types of registered "sdk" FatFramework.
```
Comment (koshachy) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:330]: "sdk" XCFramework?

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 26]
```diff
@@ -0,0 +1,335 @@
...
+ 
+To declare final native binaries, use the new binaries format with the `kotlinArtifacts` DSL. It represents a collection
+of native binaries built for this target in addition to the default `*.klib` artifact and provides a set of methods for declaring and configuring them.
+ 
+> The `kotlin-multiplatform` plugin doesn't create any production binaries by default. The only binary available by default 
+> is a debug test executable that lets you run unit tests from the `test` compilation.
+>
+{type="note"}
+
+Kotlin artifact DSL can help you solve a common issue, when your project has multiple Kotlin modules, and you need to
```
Comment (koshachy) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:19]: "issue: when your project" — this works for me better, but it's up to you.

Comment (koshachy) [reply docs/topics/multiplatform/multiplatform-native-artifacts.md:19]: Kotlin artifact DSL can help you to solve a common issue: when you need to access multiple Kotlin modules from your app. 
Does this work?

File: docs/topics/multiplatform/multiplatform-native-artifacts.md [hunk 27]
```diff
@@ -0,0 +1,335 @@
...
+modules into a single artifact with new DSL.
+
+## Declare binaries
+
+The `kotlinArtifacts` element is the top-level block for artifact configuration in the Gradle build script. Use
+the following kinds of binaries to declare elements of the `kotlinArtifacts` DSL: 
+
+| Factory method | Binary kind             | Available for                                |
+|----------------|-------------------------|----------------------------------------------|
+| `sharedLib`    | Shared native library   | All native targets, except for `WebAssembly` |
```
Comment (koshachy) [inline docs/topics/multiplatform/multiplatform-native-artifacts.md:30]: Maybe some links on what is "shared" and what is "static"? 🤔

**Review by koshachy — CHANGES_REQUESTED** [review]
Some minor comments.

**Review by koshachy — APPROVED** [review]
Niiiiice!

**Pull request merged**